Add Duration.prototype.total method

justingrant opened the pull request:
This PR fixes #584 by:
* Adding a new `total()` method on the Duration type
* Updating the `RoundDuration` abstract operation to return a remainder, which is used to calculate the fractional value returned by `total`. Other uses of `RoundDuration` will ignore the remainder.

BTW, this PR encountered problems with `RoundDuration` that I expect will be fixed by the fix to #1059. Therefore, a few of the tests in this PR are failing, but I expect them to pass after #1059's fix lands. See https://github.com/tc39/proposal-temporal/issues/1059#issuecomment-717448093 for more info on the failing tests. 

I'm very new to writing spec text so I assume I made a few mistakes in the spec. 

Questions / Open Issues: 
1. The changes to RoundDuration noted above are only one possible way to solve this problem. Would any of the alternatives below be better?
   * 1.1 Only return the remainder if an optional boolean parameter is `true`, thereby removing the need for any floating-point math in cases where rounding is being used outside of `total`.
   * 1.2 Add a boolean parameter, but if it's true then simply return a primitive floating-point value with the whole+fractional value. An advantage of doing this would be to remove some processing in cases where the resulting Duration object is not needed. I opted against this choice because I wasn't sure if it'd be weird if an abstract operation sometimes returned an object and sometimes returned a primitive. We'd never do this for a public-facing API, but I wasn't sure if it matters as much for AOs.  Also, I wasn't sure if there would be cases in the future where code would want *both* the fractional remainder *and* the Duration result. 

2. I re-used the `ToSmallestTemporalDurationUnit` and simply created a new `options` object: `{ smallestUnit: unit}`. This seems a little hacky. Is there a better way to do it?  Should this AO be renamed to  `ToTemporalDurationRoundingUnit` and accept a parameter to determine whether the name of the option is `smallestUnit` or `unit`?  And if so, should the parameter be a boolean or the string name of the unit?  Alternatively, if the current approach is correct, then how should "created a new `options` object: `{ smallestUnit: unit}`"  be expressed in spec text?  I put in a placeholder with a TODO. 

3. `UnbalanceDurationRelative` only goes down to days. If the unit is smaller, then I needed to convert days down to that unit before rounding the result. Is there already an abstract operation that does this that could be used instead of one-off code in this PR?

ptomato commented:
#1065 is ready if you want to rebase this on top of that one.

justingrant commented on docs/duration.md:
```diff
@@ -486,6 +486,58 @@ quarters = d.months / 3;
+
+Calculates the number of units of time that can fit in a particular `Temporal.Duration`.
+If the duration is not evenly divisible by the desired unit, then a fractional remainder will be included.
+If not, then the result is an integer.
```
Yeah, I struggled with this language.  Here's a revised version that I think works better. Thanks!

> If the duration IS NOT evenly divisible by the desired unit, then a fractional remainder will be present in the result.
> If the duration IS evenly divisible by the desired unit, then the integer result will be identical to `duration.round({ smallestUnit: unit, largestUnit: unit, relativeTo })[unit]`.

justingrant commented on polyfill/lib/ecmascript.mjs:
```diff
@@ -2444,6 +2445,7 @@ export const ES = ObjectAssign({}, ES2020, {
           oneYearDays = calendar.daysInYear(relativeTo);
         }
         years += days / oneYearDays;
+        remainder = years % 1;
```
Good idea. Done.  AFAICT, the existing code would have been correct for other rounding *modes*, but not for other rounding *increments*. Your version does both. 👍

justingrant commented on polyfill/test/duration.mjs:
```diff
@@ -1065,6 +1065,147 @@ describe('Duration', () => {
+    it('accepts datetime string equivalents or fields for relativeTo', () => {
+      ['2020-01-01', '2020-01-01T00:00:00.000000000', 20200101, 20200101n, { year: 2020, month: 1, day: 1 }].forEach(
+        (relativeTo) => {
+          const end = Temporal.DateTime.from(relativeTo).add(d);
```
Yeah, this was kind of intentional. I was concerned about spurious test failures due to minor differences in precision in how large numbers of nanoseconds were calculated. Given that I didn't have a working version of rounding to compare to, I didn't know if this was a worry or not.  So I assumed that by using DateTime, I'd be getting exactly the same precision calculations (at least at the msecs level) that `total` would also be using. 

Regardless, I agree with you that it's better to isolate them and I've changed my code. After your PR lands I can always come back and add rounding at the Nth decimal place to remove precision as an issue in these tests.

justingrant replied:
BTW, getting this right was more complicated than your sample above. 5 weeks is > 1 month, so you need to find the number of remainder days. For this reason, the divisor is 31 (July) not 30 (June).  For the same reason, `partialDays` needs to be adjusted.

 Here's what I ended up with:
```js
const daysPastJuly1 = 5 * 7 + 5 - 30; // 5 weeks + 5 days - 30 days in June
const partialDayNanos = ((((5 * 1e3 + 5) * 1e3 + 5) * 1e3 + 5) * 60 + 5) * 60;
const partialDay = partialDayNanos / (3.6e12 * 24);
const partialMonth = (daysPastJuly1 + partialDay) / 31;
const totalMonths = 5 * 12 + 5 + 1 + partialMonth; // +1 for 5 weeks
const total = d.total({ unit: 'months', relativeTo });
equal(total, totalMonths); // 66.32930780242619
```

justingrant replied:
> ```js
> const partialDayNanos = ((((5 * 1e3 + 5) * 1e3 + 5) * 1e3 + 5) * 60 + 5) * 60;
> ```

Actually, this code has a bug somewhere. Not sure where, but I replaced with this: 

```js
          const partialDayNanos =
            d.hours * 3.6e12 +
            d.minutes * 6e10 +
            d.seconds * 1e9 +
            d.milliseconds * 1e6 +
            d.microseconds * 1e3 +
            d.nanoseconds;
```

justingrant commented on polyfill/test/duration.mjs:
```diff
@@ -1065,6 +1065,147 @@ describe('Duration', () => {
+      throws(() => d.total({ unit: 'microseconds' }), RangeError);
+      throws(() => d.total({ unit: 'nanoseconds' }), RangeError);
+    });
+    const d2Nanoseconds = d2.round({ largestUnit: 'nanoseconds', smallestUnit: 'nanoseconds' }).nanoseconds;
```
Agree. Your expression has a bug somewhere but after spending 2 minutes trying to find it, I gave up and just did this: 
```
    const d2Nanoseconds =
      d2.days * 24 * 3.6e12 +
      d2.hours * 3.6e12 +
      d2.minutes * 6e10 +
      d2.seconds * 1e9 +
      d2.milliseconds * 1e6 +
      d2.microseconds * 1e3 +
      d2.nanoseconds;
```

ptomato replied:
Works just as well. Shooting math expressions from the hip this afternoon :stuck_out_tongue_closed_eyes:

justingrant commented on polyfill/test/duration.mjs:
```diff
@@ -1065,6 +1065,147 @@ describe('Duration', () => {
+    });
+    const endpoint = relativeTo.add(d);
+    const options = (unit) => ({ largestUnit: unit, smallestUnit: unit, roundingMode: 'trunc' });
+    const fullYears = 6;
```
Good catch. I must have mistaken years for months, whose value is actually `6`.  What's funny is that I've probably stepped through this code in the debugger many times without noticing the minus signs. It's like the [gorilla and the basketball](http://www.theinvisiblegorilla.com/gorilla_experiment.html).

justingrant commented on polyfill/test/duration.mjs:
```diff
@@ -1065,6 +1065,147 @@ describe('Duration', () => {
+    const fullMilliseconds = endpoint.since(relativeTo, options('milliseconds')).milliseconds;
+    const partialDayMilliseconds = fullMilliseconds - fullDays * 24 * 3.6e6 + 0.005005;
+    const fractionalDay = partialDayMilliseconds / (24 * 3.6e6);
+    const partialYearDays = fullDays - fullYears * 365 + 2;
```
Another good catch, thanks. Prettier is very opinionated about where you're allowed to put unnecessary parentheses, and after a while I found myself avoiding parentheses that Prettier might complain about. This might have been one of those cases. ;-(

justingrant commented on polyfill/test/duration.mjs:
```diff
@@ -1065,6 +1065,147 @@ describe('Duration', () => {
+    const partialDayMilliseconds = fullMilliseconds - fullDays * 24 * 3.6e6 + 0.005005;
+    const fractionalDay = partialDayMilliseconds / (24 * 3.6e6);
+    const partialYearDays = fullDays - fullYears * 365 + 2;
+    const fractionalYear = partialYearDays / 365 + fractionalDay / 365;
```
This was intentional. I split these out to avoid any precision loss from the division. I added a comment to explain.

justingrant commented on polyfill/test/duration.mjs:
```diff
@@ -1065,6 +1065,147 @@ describe('Duration', () => {
+    it('balances differently depending on relativeTo', () => {
+      const fortyDays = Duration.from({ days: 40 });
+      equal(fortyDays.total({ unit: 'months', relativeTo: '2020-01-01' }), 1 + 9 / 29);
+      equal(fortyDays.total({ unit: 'months', relativeTo: '2020-02-01' }), 1 + 9 / 31);
```
Yep good catch. BTW, this code never actually ran in my tests (that's why I didn't catch it before) because the previous line ran into the RoundDuration backwards/forwards bug.

justingrant replied:
> I added two assertions to the corresponding test for round() for -40 days relative to 2020-01-01 and 2020-02-01, they should both be -1 - 9/31, might be good to add them here?

Good idea. I broke them out into separate `it()` blocks to clearly differentiate negative tests from positive ones. Along those lines, I'm going to add a few more negative-duration tests too in the next checkin.

justingrant commented on spec/duration.html:
```diff
@@ -417,6 +417,52 @@ <h1>Temporal.Duration.prototype.round ( _options_ )</h1>
+        1. Perform ? RequireInternalSlot(_duration_, [[InitializedTemporalDuration]]).
+        1. Set _options_ to ? NormalizeOptionsObject(_options_).
+        1. Let _relativeTo_ be ? ToRelativeTemporalObject(_options_).
+        1. If _unit_ is not given, then
```
Yep, it's changed a lot in the latest rev.

justingrant commented on spec/duration.html:
```diff
@@ -417,6 +417,52 @@ <h1>Temporal.Duration.prototype.round ( _options_ )</h1>
+        1. Let _roundResult_ be ? RoundDuration(_unbalanceResult_.[[Years]], _unbalanceResult_.[[Months]],
+        _unbalanceResult_.[[Weeks]], _unbalanceResult_.[[Days]], _hours_, _minutes_, _seconds_, _milliseconds_,
+        _microseconds_, _nanoseconds_, _roundingIncrement_, _unit_, _roundingMode_, _relativeTo_).
+        1. Let _whole_ be ? Get(_roundResult_, _unit_).
```
> You might have to resort to something tedious like "If _unit_ is _"years"_, then / Let _whole_ be _roundResult_.[[Years]]".

OK I'll do that.

justingrant commented on polyfill/test/duration.mjs:
```diff
@@ -1065,6 +1065,147 @@ describe('Duration', () => {
+    const endpoint = relativeTo.add(d);
+    const options = (unit) => ({ largestUnit: unit, smallestUnit: unit, roundingMode: 'trunc' });
+    const fullYears = 6;
+    const fullDays = endpoint.since(relativeTo, options('days')).days;
```
If OK with you, I'd prefer to leave this one as-is. Then we'll have some tests that are using hand-coded numbers and other tests that are validating that results match up with other Temporal objects, like an integration test. But I'm also OK to change it if you feel strongly. What do you think?

justingrant commented:
> Thanks, this is great. With the two changes that I marked, the tests all pass when I rebase it on top of my #1059 branch.
> 
> Did you add it to the TypeScript file?

Haha, I'm always hassling everyone for forgetting the TS types and now I did it myself! Fixed in latest commit. 

> 
> I'll try to answer your open questions below. Other than what's below, I only had a few minor comments.
> 
> > The changes to RoundDuration noted above are only one possible way to solve this problem. Would any of the alternatives below be better?
> 
> I think this is fine. The way I was picturing doing it if I had to do it myself, would be to move the rounding out of RoundDuration altogether and always returning a fractional value (e.g. `{ years: 5, months: 5, weeks: 5, days: 5.211863484 }` or whatever) and then round it in the caller. But this works just as well and appears to require less of a change.
> 
> > I re-used the ToSmallestTemporalDurationUnit and simply created a new options object: { smallestUnit: unit}. This seems a little hacky. Is there a better way to do it?
> 
> I think we should probably refactor this into ToSingularTemporalUnit and ToPluralTemporalUnit which take the option name as a parameter, and write the current set of abstract ops in terms of those. But it's fine to make that out of scope for this PR. For now, I'd just copy the existing abstract op, change the name of the option, and call it something like ToTemporalDurationTotalUnit.

Sounds good. Done. It turned out that ToSmallestTemporalDurationUnit had more functionality (excluded strings, fallback) than this one needed so I simplified the total version of the AO and left a comment about how it's similar or different from ToSmallestTemporalDurationUnit. 
 
> > UnbalanceDurationRelative only goes down to days. If the unit is smaller, then I needed to convert days down to that unit before rounding the result. Is there already an abstract operation that does this that could be used instead of one-off code in this PR?
> 
> I believe BalanceDuration should already do this (it both balances and unbalances hours and smaller units).

Yep, that worked. Thanks!

> #1065 is ready if you want to rebase this on top of that one.

This worked!  I had two small bugs in my test code, but after I fixed those then all tests passed!  Thanks for your help with reviewing this PR and for the quick turnaround on #1065 which was really helpful to get this working!

Anyway, if you want to review the changes I made, I rolled all post-code-review changes into one commit so you can see whats changed.

justingrant commented:
Also, @ptomato once you've reviewed my changes in response to your review, should I squash this PR into one commit? This may make it easier to rebase above or below #1065 because the final commit of this PR has fewer merge conflicts with #1065 than the first commit did.

ptomato reviewed (approved):
Looks fine to me now.

> once you've reviewed my changes in response to your review, should I squash this PR into one commit?

Yes, I would recommend that.

Ms2ger commented:
Tests are failing in CI.

justingrant commented:
> Tests are failing in CI.

Yep, expected. From the OP: 

> BTW, this PR encountered problems with `RoundDuration` that I expect will be fixed by the fix to #1059. Therefore, a few of the tests in this PR are failing, but I expect them to pass after #1059's fix lands. See [#1059 (comment)](https://github.com/tc39/proposal-temporal/issues/1059#issuecomment-717448093) for more info on the failing tests.

Is #1065 ready?  If so, then you should be able to merge it, and then this PR's tests should pass.

Ms2ger commented:
Thanks, missed that comment.
